src/devices/src/virtio/mmio.rs

```diff
@@ -343,542 +343,575 @@ impl BusDevice for MmioTransport {
+            }
+        }
+    }
+
+    impl VirtioDevice for DummyDevice {
+        fn device_type(&self) -> u32 {
+            123
+        }
+
+        fn read_config(&self, offset: u64, data: &mut [u8]) {
+            data.copy_from_slice(&self.config_bytes[offset as usize..]);
+        }
+
+        fn write_config(&mut self, offset: u64, data: &[u8]) {
+            for (i, item) in data.iter().enumerate() {
+                self.config_bytes[offset as usize + i] = *item;
+            }
+        }
+
+        fn avail_features(&self) -> u64 {
+            self.avail_features
+        }
+
+        fn acked_features(&self) -> u64 {
+            self.acked_features
+        }
+
+        fn set_acked_features(&mut self, acked_features: u64) {
+            self.acked_features = acked_features;
+        }
+
+        fn activate(&mut self, _mem: GuestMemory) -> ActivateResult {
+            Ok(())
+        }
+
+        fn get_queues(&mut self) -> &mut Vec<Queue> {
+            &mut self.queues
+        }
+
+        fn get_queue_events(&self) -> result::Result<Vec<EventFd>, std::io::Error> {
+            self.queue_evts.iter().map(|evt| evt.try_clone()).collect()
+        }
+
+        fn get_interrupt(&self) -> result::Result<EventFd, std::io::Error> {
+            self.interrupt_evt.try_clone()
+        }
+
+        fn get_interrupt_status(&self) -> Arc<AtomicUsize> {
+            self.interrupt_status.clone()
+        }
+
+        fn is_activated(&self) -> bool {
+            self.device_activated
+        }
+
+        fn set_device_activated(&mut self, device_activated: bool) {
+            self.device_activated = device_activated;
+        }
+
+        fn set_avail_features(&mut self, value: u64) {
+            self.avail_features = value;
+        }
+    }
+
+    fn set_device_status(d: &mut MmioTransport, status: u32) {
+        let mut buf = vec![0; 4];
+        LittleEndian::write_u32(&mut buf[..], status);
+        d.write(0x70, &buf[..]);
+    }
+
+    #[test]
+    fn test_new() {
+        let m = GuestMemory::new(&[(GuestAddress(0), 0x1000)]).unwrap();
+        let mut dummy = DummyDevice::new();
+        // Validate reset is no-op.
+        assert!(dummy.reset().is_none());
+        let mut d = MmioTransport::new(m, Arc::new(Mutex::new(dummy))).unwrap();
+
+        // We just make sure here that the implementation of a mmio device behaves as we expect,
+        // given a known virtio device implementation (the dummy device).
+
+        assert_eq!(d.locked_device().get_queue_events().unwrap().len(), 2);
+
+        assert!(d.locked_device().get_interrupt().is_ok());
+
+        assert!(!d.are_queues_valid());
+
+        set_device_status(&mut d, device_status::ACKNOWLEDGE);
+        set_device_status(&mut d, device_status::ACKNOWLEDGE | device_status::DRIVER);
+        set_device_status(
+            &mut d,
+            device_status::ACKNOWLEDGE | device_status::DRIVER | device_status::FEATURES_OK,
+        );
```

Comment:
Setting the device status is not needed for the scope of this test because the status is only needed when doing writes on the bus.